server/infinispan/src/main/java/org/hawkular/apm/server/infinispan/InfinispanSpanCache.java

```diff
@@ -146,4 +142,25 @@ public void store(String tenantId, List<Span> spans, Function<Span, String> cach
                 .filter(span -> !span.serverSpan())
                 .collect(Collectors.toList());
     }
+
+    @Override
+    public List<Span> getTrace(String tenant, String id) {
+        if (id == null) {
+            throw new NullPointerException("Id should not be null!");
+        }
+
+        Set<Span> trace = traceCache.get(id);
```

Comment:
if returning a list (e.g. for more efficient iteration) then wouldn't it be better to cache the list - rather than creating list + initialising each time the trace is returned?